src/main/java/jnr/enxio/channels/AbstractNativeSocketChannel.java

```diff
@@ -27,102 +27,134 @@
-
-        buffer.put(src);
-
-        buffer.position(0);
-
-        int n = Native.write(fd, buffer);
-
-        if (n < 0) {
-            throw new IOException(Native.getLastErrorString());
-        }
-
-        return n;
-    }
-
-    public SocketChannel shutdownInput() throws IOException {
-        int n = Native.shutdown(fd, SHUT_RD);
-        if (n < 0) {
-            throw new IOException(Native.getLastErrorString());
-        }
-        return this;
-    }
-
-    public SocketChannel shutdownOutput() throws IOException {
-        int n = Native.shutdown(fd, SHUT_WR);
-        if (n < 0) {
-            throw new IOException(Native.getLastErrorString());
-        }
-        return this;
-    }
-
-    private static final int SHUT_RD = Shutdown.SHUT_RD.intValue();
-    private static final int SHUT_WR = Shutdown.SHUT_WR.intValue();
+public abstract class AbstractNativeSocketChannel extends SocketChannel
+		implements ByteChannel, NativeSelectableChannel {
+
+	private int fd = -1;
+
+	public AbstractNativeSocketChannel(int fd) {
+		this(NativeSelectorProvider.getInstance(), fd);
+	}
+
+	AbstractNativeSocketChannel(SelectorProvider provider, int fd) {
+		super(provider);
+		this.fd = fd;
+	}
+
+	public void setFD(int fd) {
+		this.fd = fd;
+	}
+
+	@Override
+	protected void implCloseSelectableChannel() throws IOException {
+		Native.close(fd);
+	}
+
+	@Override
+	protected void implConfigureBlocking(boolean block) throws IOException {
+		Native.setBlocking(fd, block);
+	}
+
+	public final int getFD() {
+		return fd;
+	}
+
+	public int read(ByteBuffer dst) throws IOException {
+
+		ByteBuffer buffer = ByteBuffer.allocate(dst.remaining());
+
+		int n = Native.read(fd, buffer);
+
+		buffer.flip();
+
+		dst.put(buffer);
+
+		switch (n) {
+		case 0:
+			return -1;
+
+		case -1:
+			Errno lastError = Native.getLastError();
+			switch (lastError) {
+			case EAGAIN:
+			case EWOULDBLOCK:
+				return 0;
+
+			default:
+				throw new IOException(Native.getLastErrorString());
+			}
+
+		default: {
+
+			return n;
+		}
+		}
+	}
+
+	@Override
+	public long read(ByteBuffer[] dsts, int offset, int length)
+			throws IOException {
+		long total = 0;
+
+		for (int i = 0; i < length; i++) {
+			ByteBuffer dst = dsts[offset + i];
+			long read = read(dst);
+			if (read == -1) {
+				return read;
+			}
+			total += read;
+		}
+
+		return total;
+	}
+
+	public int write(ByteBuffer src) throws IOException {
+
+		ByteBuffer buffer = ByteBuffer.allocate(src.remaining());
+
+		buffer.put(src);
+
+		buffer.position(0);
+
+		int n = Native.write(fd, buffer);
+
+		if (n < 0) {
+			throw new IOException(Native.getLastErrorString());
+		}
+
+		return n;
+	}
+
+	@Override
+	public long write(ByteBuffer[] srcs, int offset, int length)
+			throws IOException {
+
+		long result = 0;
+		int index = 0;
+
+		for (index = offset; index < length; index++) {
+			result += write(srcs[index]);
+		}
+
+		return result;
+	}
+
+	public SocketChannel shutdownInput() throws IOException {
```

Comment:
1. Could you pleas move shutdownInput() and shutdownOutput() out of the base class.
  Reason:
  Those are SOCK_STREAM specific.
  I am currently working on Datagram channels in my personal fork and would like do inherit that from AbstractNativeSocketChannel.

2. For completeness, bind(), send() and receive() should probably be moved here too. (From my UnixDatagramChannel) see my personal branch at https://github.com/felfert/jnr-unixsocket/tree/datagram 

3. I really think we should have some sort of design meeting (virtually). Have you seen my PM?